### PR TITLE
Add Npc Stats dumper

### PR DESCRIPTION
**Pull request by TheStonedTurtle**
Scraps the wiki for the below stats, if they exist, for all attackable npcs in the cache. Should handle variants as well

The dumped file will be written to `runelite/runelite-client/src/main/resources/npc_stats.json`

### Dumped stats
* hitpoints
* combat level
* slayer level
* attack level
* strength level
* defence level
* range level
* magic level
* stab bonus
* slash bonus
* crush bonus
* range bonus
* magic bonus
* stab defence bonus
* slash defence bonus
* crush defence bonus
* range defence bonus
* magic defence bonus
* bonus attack
* bonus strength
* bonus range strength
* bonus magic damage
* poison immune
* venom immune
* is a dragon
* is a demon
* is undead

### CODE REVIEW

File: src/main/java/net/runelite/data/dump/wiki/NpcStatsDumper.java [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+	{
+		final File out = new File("runelite/runelite-client/src/main/resources/");
+		out.mkdirs();
+
+		log.info("Dumping npc stats to {}", out);
+
+		final NpcManager npcManager = new NpcManager(store);
+		npcManager.load();
+
+		final Set<Integer> npcIds = getAttackableNpcIds(npcManager);
```
Comment (deathbeam) [inline src/main/java/net/runelite/data/dump/wiki/NpcStatsDumper.java]: use parallelStream like other dumpers do

File: src/main/java/net/runelite/data/dump/wiki/NpcStatsDumper.java [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+		final Set<Integer> foundIds = new HashSet<>();
+
+		log.info("Dumping {} npc stats...this may take a while", npcIds.size());
```
Comment (deathbeam) [inline src/main/java/net/runelite/data/dump/wiki/NpcStatsDumper.java]: this may take a while... is not necessary, just use exactly same format as other dumpers do

File: src/main/java/net/runelite/data/dump/wiki/NpcStatsDumper.java [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+			String wikiIdString = getWikiIdString(base, variantKey);
+			if (wikiIdString == null)
+			{
+				// Try again as `id` will be null if there are variants and `id1` is the starting key
+				variantKey++;
+				wikiIdString = getWikiIdString(base, variantKey);
+				if (wikiIdString == null)
+				{
+					log.warn("The wiki page for npc `{}` does not have an `id` or `id1` value");
+					continue;
```
Comment (deathbeam) [inline src/main/java/net/runelite/data/dump/wiki/NpcStatsDumper.java]: this is not necessary

File: src/main/java/net/runelite/data/dump/wiki/NpcStatsDumper.java [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+
+		try (FileWriter fw = new FileWriter(new File(out, "npc_stats.json")))
+		{
+			fw.write(App.GSON.toJson(npcStats));
+		}
+
+		log.info("Dumped {} npc stats", npcStats.size());
+
+		Set<Integer> missed = Sets.difference(npcIds, foundIds);
+		log.info("Missed {} npcs", missed.size());
```
Comment (deathbeam) [inline src/main/java/net/runelite/data/dump/wiki/NpcStatsDumper.java]: not necessary

**Review by deathbeam — CHANGES_REQUESTED** [review]

**Comment by TheStonedTurtle** [comment]
The PR has been updated to use parallel streams as well as fix a few other issues.

The first two commits are fixing "issues" that exist in the client that i ran into. I also ran into an issue with the wiki having comments in the `getInt()` value causing errors with the parseInt. I fixed this in my custom getInt() as i wasn't sure if updating the main method was a good idea.

TL;DR needs a full review again :)

File: src/main/java/net/runelite/data/dump/MediaWikiTemplate.java [hunk 5]
```diff
@@ -99,8 +101,11 @@
 	@Nullable
 	public static MediaWikiTemplate parseWikitext(final String name, final String data)
 	{
+		final Pattern exactNameTest = Pattern.compile(name.toLowerCase() + "\\s*\\|");
```
Comment (deathbeam) [inline src/main/java/net/runelite/data/dump/MediaWikiTemplate.java]: you can just make the pattern case-insensitive. Also, if you want to match | also match {{ at start

**Review by deathbeam — CHANGES_REQUESTED** [review]

**Comment by TheStonedTurtle** [comment]
Okay i think this is ready again
